### PR TITLE
client/core: resubscribe to order books

### LINKED ISSUE #463: core: need to resubscribe to order books on reconnect
Should be done in `(*Core).handleReconnect`. 

The `BookRouter` effectively associates the subscription with the `comms.Link`, so subscriptions must be renewed on reconnection.

### PR DESCRIPTION
**Pull request by song50119**
Resolve #463

### CODE REVIEW

File: client/core/core.go [hunk 1]
```diff
@@ -2392,6 +2392,44 @@ func (c *Core) handleReconnect(host string) {
...
+			continue
+		}
+
+		booky := newBookie(func() { c.unsub(dc, mktKey) })
+		err = booky.Sync(result)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		dc.books[mktKey] = booky
```
Comment (JoeGruffins) [inline client/core/core.go:2430]: I think the `dc.booksMtx` lock can be held just for this one action, rather than the whole loop.

Comment (song50119) [reply client/core/core.go:2430]: Okay

**Comment by song50119** [comment]
@JoeGruffins 
Can you reproduce to get `2020-06-23 18:36:59.532 [ERR] CORE: error unsubscribing from dcr_btc` 
I think these logs implemented on other PRs `2020-06-23 18:41:20.003 [TRC] CORE: notify: |DATA| (epoch) - Index: 159290528`
Thank you

**Comment by dnldd** [comment]
Looks good, haven't tested yet however.

File: client/core/types.go [hunk 2]
```diff
@@ -151,6 +151,7 @@ type Market struct {
 	Orders          []*Order `json:"orders"`
 	pendingSuspend  *time.Timer
 	suspended       bool
+	subscribed      bool
```
Comment (buck54321) [inline client/core/types.go:154]: Using this new field is unnecessary. If there is an entry in the `(*dexConnection).books` map, the client is subscribed.

File: client/core/core.go [hunk 3]
```diff
@@ -2392,6 +2392,47 @@ func (c *Core) handleReconnect(host string) {
 		log.Errorf("unable to authorize DEX at %s: %v", host, err)
 		return
 	}
+
+	dc.marketMtx.RLock()
+	for mktKey, mkt := range dc.marketMap {
+		if !mkt.Subscribed() {
```
Comment (buck54321) [inline client/core/core.go:2398]: Just loop through the `(*dexConnection).books` map to see what subscriptions are open.

File: client/core/core.go [hunk 4]
```diff
@@ -2392,6 +2392,50 @@ func (c *Core) handleReconnect(host string) {
...
+		}
+
+		booky := newBookie(func() { c.unsub(dc, mktKey) })
+		err = booky.Sync(result)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		dc.booksMtx.Lock()
+		dc.books[mktKey] = booky
```
Comment (buck54321) [inline client/core/core.go:2435]: One issue is that the e.g. webserver will still be listening on the old `bookie`'s `BookFeed`. We need a way to either 1) Pass the old `bookie`'s `BookFeed`s to the new `bookie`, or 2) notify the caller that the old feed is dead, possibly by closing the channel. See `marketSyncer` too.

I actually think solution 1 is the best.

File: client/core/core.go [hunk 5]
```diff
@@ -2392,6 +2392,50 @@ func (c *Core) handleReconnect(host string) {
...
+		dc.booksMtx.Unlock()
+		if !foundMkt {
+			continue
+		}
+		req, err := msgjson.NewRequest(dc.NextID(), msgjson.OrderBookRoute, &msgjson.OrderBookSubscription{
+			Base:  mkt.BaseID,
+			Quote: mkt.QuoteID,
+		})
+		if err != nil {
+			log.Errorf("error encoding 'orderbook' request: %v", err)
```
Comment (buck54321) [inline client/core/core.go:2409]: Can you call `(*Core).Sync` here instead?

**Pull request merged**